src/uri.rs

```diff
@@ -0,0 +1,87 @@
+use std::{hash::Hash, ops::Deref, str::FromStr};
+
+use serde::{de::Error, Deserialize, Serialize};
+
+/// Newtype struct around `url::Url` with serialization implementations that properly encode brackets for LSP compatibility.
+#[derive(Debug, Clone)]
+pub struct Uri(url::Url);
```

Comment:
```suggestion
pub struct Uri(pub url::Url);
```

One concern I have is how Zed deals with this: if there's a need to construct the struct in Zed, there's no good way now but through the parsing.